### ISSUE TITLE
looks like Pipelines Controller is not freeing memory

### ISSUE THREAD

**Issue by jhutar** [tektoncd/pipeline#7691]
# Expected Behavior

I would expect after some time Pipelines Controlled under load memory consumption would become constant - Pipelines Controller will start freeing memory.

# Actual Behavior

This is a memory graph of a Pipelines Controller processing 10k very simple PipelineRuns with just one Task just printing "hello world" ([Pipeline](https://github.com/openshift-pipelines/performance/blob/d3ac7103bbd1a52375daf5a2fb8429526c9a741d/tests/scaling-pipelines/scenario/signing-tr-tekton-bigbang/pipeline.yaml), [PipelineRun](https://github.com/openshift-pipelines/performance/blob/d3ac7103bbd1a52375daf5a2fb8429526c9a741d/tests/scaling-pipelines/scenario/signing-tr-tekton-bigbang/run.yaml))

These PRs were running from about 13:00 to to 15:30, script was creating new PRs in a way to make sure at max 100 of them is in pending/running:

![image](https://github.com/tektoncd/pipeline/assets/3397250/78b9956e-296e-449c-936f-ddc7fe2c9757)

Is this expected, or is this some sort of memory leak?

# Steps to Reproduce the Problem

1. Run 10k PipelineRuns and observe Pipelines Controller memory consumption over time
2. This was automated in this [repo](https://github.com/openshift-pipelines/performance/blob/d3ac7103bbd1a52375daf5a2fb8429526c9a741d/README.md) with `signing-tr-tekton-bigbang` scenario

# Additional Info

- Kubernetes version:

Cluster is gone already, but it was ROSA OpenShift 4.14.11 with 5 compute nodes AWS EC2 m6a.2xlarge

- Tekton Pipeline version:

* chains-info: v0.19.0
* pipelines-as-code-info: v0.22.4
* pipelines-info: v0.53.2
* triggers-info: v0.25.3
* openshift-pipelines-operator-cluster-operations: v0.69.0
* openshift-pipelines-operator-lifecycle: 1.13.0

Reported this together with https://github.com/tektoncd/chains/issues/1058

**Comment by vdemeester** [comment]
@jhutar are the PipelineRun cleaned (deleted) from the cluster in that scenario ? Because if not, they will be "cached" by the informers and thus, they will be kept in memory up until they are deleted from the cluster (and informers cache updated).

**Comment by jhutar** [comment]
Yes, they are not deleted. So this means Pipelines Controller memory usage will just grow with number and size of PRs in the cluster. If there are excessively big PRs (maybe with long script), this might become a problem even for smaller number of PRs.

I understand this is property of underlying Go library (these informers you mentioned), but is there a way how to drop oldest records from the cache or so? Just to have a way how to keep memory usage flat.

**Comment by vdemeester** [comment]
> I understand this is property of underlying Go library (these informers you mentioned), but is there a way how to drop oldest records from the cache or so? Just to have a way how to keep memory usage flat.

We need to explore this yes. There might be ways to "optimize" or filter some objects once they are done.

**Comment by sibelius** [comment]
any way to avoid this ?

**Comment by vdemeester** [comment]
## Investigation Summary

I've investigated the memory leak issue and identified the root cause along with a proven solution from the Kubernetes community.

### Root Cause

The controller uses `SharedInformerFactory` which caches **all PipelineRun objects indefinitely** without any eviction policy. While event filtering prevents unnecessary reconciliation triggers, it does **not** prevent caching - all PipelineRuns remain in memory.

**Key findings:**

1. **Informer factory** (`pkg/client/injection/informers/factory/factory.go:45`) only configures namespace scope - no cache optimization
2. **Event filtering** (`pkg/reconciler/pipelinerun/controller.go:112-124`) filters reconciliation events but objects still get cached
3. **Completed PipelineRuns** are reconciled once for cleanup (AffinityAssistants, PVCs) then never touched again, but remain cached forever

**Memory-heavy fields** in PipelineRunStatus:
- `PipelineSpec` - entire pipeline definition embedded (~10-50 KB)
- `Provenance` - build provenance metadata
- `Results` - execution results (potentially large)
- `ChildReferences` - all TaskRuns/CustomRuns created
- Plus: `managedFields` and `kubectl.kubernetes.io/last-applied-configuration` in metadata

### Proposed Solution: WithTransform

Add a transform function to strip unnecessary fields before caching, following proven patterns from:

1. **[Contour Ingress Controller](https://github.com/projectcontour/contour/pull/5099)** - Used `WithTransform` to strip `Secret.data`, achieved **50x memory reduction**
2. **[KubeVela](https://github.com/kubevela/kubevela/pull/5683)** - Stripped `managedFields` + deduplication, achieved **25-30% reduction**

### Implementation Approach

**Phase 1: Metadata Stripping (Quick Win)**

Strip fields used only by kubectl/API server:
- `managedFields` (Server-Side Apply tracking)
- `kubectl.kubernetes.io/last-applied-configuration` annotation

Expected impact: **25-30% memory reduction** (based on KubeVela results)

**Phase 2: Status Field Stripping (Completed PipelineRuns)**

For completed PipelineRuns only, strip:
- `PipelineSpec` (embedded definition)
- `Provenance` (build metadata)
- `SpanContext` (tracing context)

Expected additional impact: **20-40% more reduction**

**Combined estimate: 40-60% memory reduction**

### Code Sketch

```go
// In pkg/client/injection/informers/factory/factory.go

func withInformerFactory(ctx context.Context) context.Context {
    c := client.Get(ctx)
    opts := make([]externalversions.SharedInformerOption, 0, 2)

    if injection.HasNamespaceScope(ctx) {
        opts = append(opts, externalversions.WithNamespace(injection.GetNamespaceScope(ctx)))
    }

    // Add transform to reduce cache memory
    opts = append(opts, externalversions.WithTransform(transformForCache))

    return context.WithValue(ctx, Key{},
        externalversions.NewSharedInformerFactoryWithOptions(c, controller.GetResyncPeriod(ctx), opts...))
}

func transformForCache(obj interface{}) (interface{}, error) {
    // Handle tombstones during deletion
    if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
        if transformed, err := transformForCache(tombstone.Obj); err == nil {
            return cache.DeletedFinalStateUnknown{
                Key: tombstone.Key,
                Obj: transformed,
            }, nil
        }
        return obj, nil
    }

    pr, ok := obj.(*v1.PipelineRun)
    if !ok {
        return obj, nil
    }

    // Phase 1: Strip metadata fields (all PipelineRuns)
    pr.ManagedFields = nil
    if pr.Annotations != nil {
        delete(pr.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
    }

    // Phase 2: Strip heavy status fields (completed only)
    if pr.IsDone() {
        pr.Status.PipelineSpec = nil
        pr.Status.Provenance = nil
        pr.Status.SpanContext = nil
    }

    return pr, nil
}
```

### Why This Approach?

✅ **Non-breaking** - Still sees all PipelineRuns, just with reduced cache size
✅ **Low risk** - Fields stripped from cache only, not from etcd
✅ **Proven pattern** - Battle-tested by Contour and KubeVela
✅ **Simple** - Minimal code changes, no architectural refactoring
✅ **Fast** - Can be implemented in 1-2 days

### Testing Plan

1. Add feature flag for gradual rollout
2. Test with 10,000 PipelineRuns scenario (matching your environment)
3. Measure actual memory reduction
4. Verify no functional regressions
5. Profile reconciliation latency impact

### Alternative: Metadata-Only Informers

For even greater reduction (~90%), we could use metadata-only informers (Kubernetes 1.19+) for completed PipelineRuns, but this requires:
- Significant refactoring (separate informers for running vs completed)
- 2-4 weeks implementation effort
- More complex migration strategy

Recommend starting with `WithTransform` as a quick win, then evaluate if further optimization is needed.

### References

- [Kubernetes Community: Mastering Informers](https://www.plural.sh/blog/manage-kubernetes-events-informers/)
- [Avoiding the Memory Pitfall](https://dev.to/jotak/kubernetes-operators-avoiding-the-memory-pitfall-10le)
- [Contour PR #5099](https://github.com/projectcontour/contour/pull/5099)
- [KubeVela PR #5683](https://github.com/kubevela/kubevela/pull/5683)

---

Happy to work on a PoC implementation if this approach looks good! Would appreciate feedback from maintainers on:
1. Is `WithTransform` acceptable for the Tekton architecture?
2. Should we include a feature flag for gradual rollout?
3. Any concerns about stripping these specific fields from cache?
